File: scripts/render_local_config.py
#!/usr/bin/env python3
import json, os, secrets
from urllib.parse import urlparse

base_url = os.environ.get('LOCALCLAW_BASE_URL', 'http://127.0.0.1:11434/v1').rstrip('/')
model_id = os.environ.get('LOCALCLAW_MODEL', 'llama3.2:latest').strip() or 'llama3.2:latest'
state_dir = os.environ.get('LOCALCLAW_STATE_DIR', os.path.join(os.getcwd(), '.localclaw', 'state'))
config_path = os.environ.get('OPENCLAW_CONFIG_PATH', os.path.join(os.getcwd(), '.localclaw', 'openclaw.local.json'))

u = urlparse(base_url)
if u.scheme not in {'http', 'https'}:
    raise SystemExit(f'LOCALCLAW_BASE_URL must be http(s), got: {base_url}')
if u.hostname not in {'127.0.0.1', 'localhost', '::1'}:
    raise SystemExit(f'LOCALCLAW refuses non-local host: {u.hostname}')

os.makedirs(os.path.dirname(config_path), exist_ok=True)
os.makedirs(state_dir, exist_ok=True)
workspace = os.path.join(state_dir, 'workspace')
os.makedirs(workspace, exist_ok=True)

def model_def(mid):
    return {
      'id': mid,
      'name': mid,
      'api': 'openai-completions',
      'input': ['text'],
      'reasoning': False,
      'cost': {'input': 0, 'output': 0, 'cacheRead': 0, 'cacheWrite': 0},
      'contextWindow': 131072,
      'maxTokens': 8192,
    }

seed_models = [
  'llama3.2:latest',
  'qwen2.5:latest',
  'mistral:latest',
  'phi4:latest',
  'gpt-oss:20b'
]
if model_id not in seed_models:
    seed_models.insert(0, model_id)

models_map = {f'openai/{m}': {'alias': m} for m in seed_models}
models_map[f'openai/{model_id}'] = {'alias': model_id}

cfg = {
  'messages': {'ackReactionScope': 'group-mentions'},
  'agents': {
    'defaults': {
      'workspace': workspace,
      'maxConcurrent': 4,
      'subagents': {'maxConcurrent': 8},
      'compaction': {'mode': 'safeguard'},
      'models': models_map,
      'model': {'primary': f'openai/{model_id}'},
    }
  },
  'gateway': {
    'mode': 'local',
    'bind': 'loopback',
    'port': int(os.environ.get('LOCALCLAW_GATEWAY_PORT', '18789')),
    'auth': {'mode': 'token', 'token': os.environ.get('LOCALCLAW_GATEWAY_TOKEN', secrets.token_hex(24))},
    'tailscale': {'mode': 'off', 'resetOnExit': False}
  },
  'models': {
    'mode': 'merge',
    'providers': {
      'openai': {
        'baseUrl': base_url,
        'api': 'openai-completions',
        'authHeader': False,
        'apiKey': 'local-only',
        'models': [model_def(m) for m in seed_models],
      }
    }
  }
}

with open(config_path, 'w', encoding='utf-8') as f:
    json.dump(cfg, f, indent=2)
print(config_path)

File: scripts/render-local-config.mjs
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';

const cwd = process.cwd();
const baseUrlRaw = (process.env.LOCALCLAW_BASE_URL ?? 'http://127.0.0.1:11434/v1').replace(/\/$/, '');
const modelId = (process.env.LOCALCLAW_MODEL ?? 'llama3.2:latest').trim() || 'llama3.2:latest';
const stateDir = process.env.LOCALCLAW_STATE_DIR ?? path.join(cwd, '.localclaw', 'state');
const configPath = process.env.OPENCLAW_CONFIG_PATH ?? path.join(cwd, '.localclaw', 'openclaw.local.json');
const tokenPath = process.env.LOCALCLAW_GATEWAY_TOKEN_PATH ?? path.join(path.dirname(configPath), 'gateway.token');

let parsed;
try {
  parsed = new URL(baseUrlRaw);
} catch {
  throw new Error(`LOCALCLAW_BASE_URL must be a valid URL, got: ${baseUrlRaw}`);
}
if (!['http:', 'https:'].includes(parsed.protocol)) {
  throw new Error(`LOCALCLAW_BASE_URL must be http(s), got: ${baseUrlRaw}`);
}
if (!['127.0.0.1', 'localhost', '::1'].includes(parsed.hostname)) {
  throw new Error(`LOCALCLAW refuses non-local host: ${parsed.hostname}`);
}

const readTextFile = (filePath) => {
  try {
    return fs.readFileSync(filePath, 'utf8').trim();
  } catch {
    return '';
  }
};

const readTokenFromConfig = () => {
  try {
    const existing = JSON.parse(fs.readFileSync(configPath, 'utf8'));
    const token = existing?.gateway?.auth?.token;
    return typeof token === 'string' ? token.trim() : '';
  } catch {
    return '';
  }
};

const isLikelyToken = (value) => typeof value === 'string' && /^[a-f0-9]{24,}$/i.test(value);
const shouldRotate = ['1', 'true', 'yes'].includes(String(process.env.LOCALCLAW_ROTATE_GATEWAY_TOKEN ?? '').toLowerCase());

const envToken = (process.env.LOCALCLAW_GATEWAY_TOKEN ?? '').trim();
const configToken = shouldRotate ? '' : readTokenFromConfig();
const fileToken = shouldRotate ? '' : readTextFile(tokenPath);
const token =
  (isLikelyToken(envToken) && envToken) ||
  (isLikelyToken(configToken) && configToken) ||
  (isLikelyToken(fileToken) && fileToken) ||
  crypto.randomBytes(24).toString('hex');

fs.mkdirSync(path.dirname(configPath), { recursive: true });
fs.mkdirSync(stateDir, { recursive: true });
const workspace = path.join(stateDir, 'workspace');
fs.mkdirSync(workspace, { recursive: true });
fs.writeFileSync(tokenPath, `${token}\n`, 'utf8');

const modelDef = (id) => ({
  id,
  name: id,
  api: 'openai-completions',
  input: ['text'],
  reasoning: false,
  cost: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0 },
  contextWindow: 131072,
  maxTokens: 8192,
});

const seedModels = ['llama3.2:latest', 'qwen2.5:latest', 'mistral:latest', 'phi4:latest', 'gpt-oss:20b'];
if (!seedModels.includes(modelId)) seedModels.unshift(modelId);

const modelsMap = Object.fromEntries(seedModels.map((m) => [`openai/${m}`, { alias: m }]));
const gatewayPort = Number.parseInt(process.env.LOCALCLAW_GATEWAY_PORT ?? '18789', 10);

const cfg = {
  messages: { ackReactionScope: 'group-mentions' },
  agents: {
    defaults: {
      workspace,
      maxConcurrent: 4,
      subagents: { maxConcurrent: 8 },
      compaction: { mode: 'safeguard' },
      models: modelsMap,
      model: { primary: `openai/${modelId}` },
    },
  },
  gateway: {
    mode: 'local',
    bind: 'loopback',
    port: Number.isFinite(gatewayPort) ? gatewayPort : 18789,
    auth: { mode: 'token', token },
    tailscale: { mode: 'off', resetOnExit: false },
  },
  models: {
    mode: 'merge',
    providers: {
      openai: {
        baseUrl: baseUrlRaw,
        api: 'openai-completions',
        authHeader: false,
        apiKey: 'local-only',
        models: seedModels.map(modelDef),
      },
    },
  },
};

fs.writeFileSync(configPath, `${JSON.stringify(cfg, null, 2)}\n`, 'utf8');
console.log(configPath);
